### ISSUE TITLE
[BUG] Maestral doesn't resume sync if it is interrupted, but rather restart from the start

### ISSUE THREAD

**Issue by raffaem** [SamSchott/maestral#289]
**Describe the bug**
Maestral shows me the syncing icon. When I click on it, it is syncing.

However every folder of my Dropbox seems up to date:

```
$ for d in ./*/ ; do (maestral file-status "$d"); done
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
up to date
```

**To Reproduce**
Just start Maestral and it will start to sync

**Expected behaviour**
If everything is up-to-date, there should be nothing to sync
**System (please complete the following information):**
 - Maestral version: GUI v. 1.3.1, daemon v. 1.3.1
 - OS: [e.g. Ubuntu] Ubuntu 20.10
 - Desktop environment: [e.g. Gnome 3.32] GNOME 3.38.1
 - Qt version:
```
$ qmake --version
QMake version 3.1
Using Qt version 5.14.2 in /usr/lib/x86_64-linux-gnu
```

**Comment by raffaem** [comment]
Hello,

it is syncing since yesterday (while everything was already syncing)

Everything seems up to date

I don't know what triggered this behavior

**Comment by SamSchott** [comment]
Does it keep syncing on a restart? It looks like it might not *actually* be syncing but just reporting a wrong status.

**Comment by raffaem** [comment]
> Does it keep syncing on a restart? It looks like it might not _actually_ be syncing but just reporting a wrong status.

Yes, it resumes syncing at restart. I don't know. How can I check what it is actually doing?

@SamSchott I need to fix this ASAP

**Comment by raffaem** [comment]
May it be due to the fact that I have uninstalled and reinstalled Maestral multiple times?

In my Dropbox folder I see no hidden folders.

Which suggests that Maestral saves its cache somewhere else.

May the cache have been removed during uninstall? 

Where can I find the cache in order to back it up?

**Comment by SamSchott** [comment]
> How can I check what it is actually doing?

The `maestral activity` command will give you a live view of what is currently uploading or downloading. If there are no paths in this list, there is nothing actually syncing.

The log will give information about past activity, e.g., if syncing got stuck somewhere. You can show it with `maestral log show`. If you want to output more detailed logs, you can set the log level to debug with `maestral log level DEBUG`.

**Comment by raffaem** [comment]
The `maestral activity` hangs. No output, but neither program termination.

Last lines of `maestral log show` are these. It's 16:43 here, so it's almost one hour that it is stuck on the last line

```
2021-01-27 15:32:28 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:32:28 maestral.client INFO: Indexing 350335...
2021-01-27 15:34:12 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:34:12 maestral.client INFO: Indexing 350835...
2021-01-27 15:35:52 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:35:52 maestral.client INFO: Indexing 351335...
2021-01-27 15:37:30 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:37:30 maestral.client INFO: Indexing 351835...
2021-01-27 15:39:24 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:39:24 maestral.client INFO: Indexing 352335...
2021-01-27 15:41:17 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:41:17 maestral.client INFO: Indexing 352835...
2021-01-27 15:43:02 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:43:02 maestral.client INFO: Indexing 353335...
2021-01-27 15:44:36 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:44:36 maestral.client INFO: Indexing 353835...
2021-01-27 15:46:26 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:46:26 maestral.client INFO: Indexing 354335...
2021-01-27 15:48:11 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:48:11 maestral.client INFO: Indexing 354835...
2021-01-27 15:49:59 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 15:49:59 maestral.client INFO: Indexing 355335...
```

**Comment by SamSchott** [comment]
Thanks for the log output! It is particularly concerning that the other CLI commands hang as well, it indicates that there may be deadlock in the main thread. This will be difficult to debug, please bear with me.

From the log output, it appears that syncing is getting stuck during the first indexing phase. When restarted, it will begin indexing from the beginning again but not repeat any downloads which were already completed. Could you answer the following questions?

1. When you stop the daemon with `maestral stop`, does it say `[KILLED]` instead of `[OK]`? This will be a clear indication that the entire process is unresponsive.
2. After stopping / killing the process, setting `maestral log level DEBUG` should succeed. Can you then start again and send me the more detailed logs? This should hopefully help to narrow down where it hangs.
3. Is the position where it hangs reproducible? Is it always at "Indexing 355335..." or sometimes earlier / later?

**Comment by raffaem** [comment]
1. The command hangs
```
$ maestral stop
Stopping Maestral...
```
never finishes.

I had to run `kill -9 63227` to kill `maestral stop`. With just `kill` it didn't got killed.

I killed other maestral processes too

2. ok
```
$ maestral log level DEBUG
✓ Log level set to DEBUG.
```

3. `maestral activity` is showing activity but it seems that it started syncing from scratch again.

Anyway tomorrow my USB key will arrive and I will reinstall Windows.

```
2021-01-27 18:28:47 maestral.sync INFO: Creating folders...
2021-01-27 18:28:47 maestral.sync DEBUG: Equal revs for "<REDACTED>": local item is the same or newer than on Dropbox
2021-01-27 18:28:47 maestral.sync INFO: Creating folder 1/1...
2021-01-27 18:28:47 maestral.sync DEBUG: Equal revs for "<REDACTED>": local item is the same or newer than on Dropbox
2021-01-27 18:28:47 maestral.sync DEBUG: Equal revs for "<REDACTED>": local item is the same or newer than on Dropbox
2021-01-27 18:28:47 maestral.client INFO: Indexing 24...
```

It doesn't update automatically but it seems not to be hanged. If I re-run the program I get new logs. (why are we using vim here?)

**Comment by SamSchott** [comment]
Regarding 1: Damn, that's not a good sign. This means that all threads are stalled -- the stop command gets sent to the daemon but then blocks.

Regarding 3: Yes, it will start indexing again from the beginning since the initial sync was aborted. But any files which have already been downloaded will be skipped ("local item is the same or newer than on Dropbox") so the sync should be much quicker. Let me know how it goes!

**Comment by raffaem** [comment]
> Regarding 3: Yes, it will start indexing again from the beginning since the initial sync was aborted. But any files which have already been downloaded will be skipped ("local item is the same or newer than on Dropbox") so the sync should be much quicker. Let me know how it goes!

The problem is that I have zillion of files each a few bytes in size. So it gets a lot of time to recognize even that

**Comment by SamSchott** [comment]
I know. This is something on my todo list. But I also have a day job!

**Comment by raffaem** [comment]
> I know. This is something on my todo list. But I also have a day job!

I know. That's why I'm gonna reinstall Windows.

Does maestral keeps the cache somewhere? I'm guessing it was removed when I uninstalled it

**Comment by SamSchott** [comment]
> Does maestral keeps the cache somewhere? I'm guessing it was removed when I uninstalled it.

It does keep a very short-lived cache inside the local Dropbox folder itself (`.maestral.cache`). More importantly, already downloaded files will remain in the "Dropbox (Maestral)" folder when uninstalling.

However, I'm still interested in finding what exactly was blocking. So if it's possible, please do keep the current sync running to see if it succeeds!

I'll also work on resuming an interrupted indexing job -- the changes should be fairly straight forward. It's just not been among my priorities because a) there have been very few reports if the initial sync failing and b) indexing is relatively fast with a "small" number of 10,000 items. One downside of not having analytics: I don't know about typical use cases and Dropbox sizes...

**Comment by raffaem** [comment]
Hello

`maestral activity` hanged again

`maestral log show` shows

```
2021-01-27 19:36:15 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 19:36:15 maestral.client INFO: Indexing 192369...
2021-01-27 19:36:49 maestral.sync INFO: Creating folders...
2021-01-27 19:36:49 maestral.sync INFO: Creating folder 1/1...
2021-01-27 19:36:49 maestral.client INFO: Indexing 192868...
2021-01-27 19:37:08 maestral.sync INFO: Syncing ↓ 1/500
2021-01-27 19:37:08 maestral.client INFO: Indexing 193368...
2021-01-27 19:37:32 maestral.sync INFO: Creating folders...
2021-01-27 19:37:32 maestral.sync INFO: Creating folder 1/16...
2021-01-27 19:37:32 maestral.client INFO: Indexing 193868...
```

but it does not seem to be hanged (it is 19:37 here)

**Comment by SamSchott** [comment]
> Not understand how it is supposed to work. If I restart my PC it will start re-indexing again?

Probably yes, because the initial indexing never completed. Once / if it completes, a "cursor" is stored in a file at `~/.local/share/maestral/maestral.state`. This cursor indicates the last synced state. When resuming the sync, Maestral will ask the Dropbox API for all changes since this cursor instead of doing another full indexing.

Uninstalling and reinstalling will leave all settings, states and databases intact. Settings are stored at `~/.config/maestral/` and sync state is stored in `~/.local/share/maestral/`. Your access token should be stored in the Gnome keyring.

**Comment by raffaem** [comment]
Hello,

1. `maestral activity` hangs. How to debug?
I have run `python3 -m trace --trace /home/raffaele/.local/bin/maestral activity >> maestral-activity-trace.log`. Do you need the file?

EDIT: Now `maestral activity` works again. Seems the hanging is random

**Comment by raffaem** [comment]
I am receiving this errors in the logs:

```
2021-01-28 06:11:58 maestral.sync INFO: Could not sync <REDACTED>
Traceback (most recent call last):
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 183, in convert_api_errors
    yield
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 942, in list_folder_iterator
    res = self.dbx.files_list_folder_continue(res.cursor)
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/base.py", line 1950, in files_list_folder_continue
    r = self.request(
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/dropbox_client.py", line 314, in request
    res = self.request_json_string_with_retry(host,
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/dropbox_client.py", line 464, in request_json_string_with_retry
    return self.request_json_string(host,
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/dropbox_client.py", line 570, in request_json_string
    raise InternalServerError(request_id, r.status_code, r.text)
dropbox.exceptions.InternalServerError: InternalServerError('f1c36efb343a40b7aa2f4d768743e437', 504, '')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/sync.py", line 2403, in get_remote_folder
    for res in list_iter:
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 953, in list_folder_iterator
    logger.info(IDLE)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 185, in convert_api_errors
    raise dropbox_to_maestral_error(exc, dbx_path, local_path)
maestral.errors.DropboxServerError: Could not sync file or folder. Something went wrong with the job on Dropbox’s end. Please verify on the Dropbox website if the job succeeded and try again if it failed.
```

```
2021-01-28 06:11:58 maestral.sync INFO: Syncing ↓ /<REDACTED>
2021-01-28 06:12:07 maestral.sync INFO: Creating folders...
2021-01-28 06:12:07 maestral.sync DEBUG: Equal revs for "/<REDACTED>": local item is the same or newer than on Dropbox
2021-01-28 06:12:07 maestral.sync INFO: Creating folder 1/1...
2021-01-28 06:12:07 maestral.client INFO: Indexing 1...
2021-01-28 06:12:53 maestral.sync INFO: Could not sync <REDACTED>
Traceback (most recent call last):
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 183, in convert_api_errors
    yield
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 942, in list_folder_iterator
    res = self.dbx.files_list_folder_continue(res.cursor)
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/base.py", line 1950, in files_list_folder_continue
    r = self.request(
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/dropbox_client.py", line 314, in request
    res = self.request_json_string_with_retry(host,
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/dropbox_client.py", line 464, in request_json_string_with_retry
    return self.request_json_string(host,
  File "/home/raffaele/.local/lib/python3.8/site-packages/dropbox/dropbox_client.py", line 570, in request_json_string
    raise InternalServerError(request_id, r.status_code, r.text)
dropbox.exceptions.InternalServerError: InternalServerError('cc65c4f2d5a74e1d929d8dae0b890079', 500, '')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/sync.py", line 2403, in get_remote_folder
    for res in list_iter:
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 953, in list_folder_iterator
    logger.info(IDLE)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/raffaele/.local/lib/python3.8/site-packages/maestral/client.py", line 185, in convert_api_errors
    raise dropbox_to_maestral_error(exc, dbx_path, local_path)
maestral.errors.DropboxServerError: Could not sync file or folder. Something went wrong with the job on Dropbox’s end. Please verify on the Dropbox website if the job succeeded and try again if it failed.
```

**Comment by raffaem** [comment]
First tit showed be this error message

![Untitled](https://user-images.githubusercontent.com/54762742/106134530-edda5780-6166-11eb-9074-737882ee42de.png)

Now the QT GUI is saying that there are no sync errors .... was the sync error fixed?

I find maestral very hard to debug ... I catched thoses exceptions by chance, maestral didn't push any exception to me

P.S. the sync seems to have finished finally ....

**Comment by SamSchott** [comment]
> P.S. the sync seems to have finished finally ....

Hurray! Dropbox servers have been undergoing scheduled maintenance yesterday, the sync error likely is a manifestation of that. If it has disappeared it means that sync was successfully completed later. There are still lingering issues with API client getting very delayed notifications for file changes on the server. Hopefully this will be resolved in the coming days.

The *should* be desktop notifications for sync issues, etc., but they depend on platform settings.

I've also started working on resuming the initial indexing seamlessly if it has been interrupted. It will still be some time until the changes are stable and can be released -- I'll let you know here when there are updates.

**Comment by raffaem** [comment]
I am not 100% sure, but I think that, after I restarted my PC, Maestral started to sync from the start again

**Comment by raffaem** [comment]
> I am not 100% sure, but I think that, after I restarted my PC, Maestral started to sync from the start again

Yes, it restarted again from the beginning.

It will never sync :(

**Comment by SamSchott** [comment]
Closing since this is fixed in 1.4.0.
